functions/Add-DbaAgListener.ps1

```diff
@@ -92,27 +102,73 @@ function Add-DbaAgListener {
             return
         }
 
+        if ($Dhcp) {
+            if (Test-Bound -ParameterName IPAddress) {
+                Stop-Function -Message "You cannot specify both an IP address and the Dhcp switch."
+                return
+            }
+
+            if ($SubnetMask.Count -gt 1) {
+                Stop-Function -Message "You can only specify a single subnet when using Dhcp."
+            }
+        }
+
         if ($SqlInstance) {
             $InputObject += Get-DbaAvailabilityGroup -SqlInstance $SqlInstance -SqlCredential $SqlCredential -AvailabilityGroup $AvailabilityGroup
         }
 
+        if (Test-Bound -ParameterName IPAddress) {
+            if ($IPAddress.Count -ne $SubnetMask.Count) {
+                if ($SubnetMask.Count -eq 1) {
+                    # If one subnet mask is supplied, let's assume we want to use the same one for every IP
+                    $SubnetMask = $SubnetMask * $IPAddress.Count
+                } else {
+                    Stop-Function -Message "When specifying multiple IP addresses, the number of subnet masks must match, or give one mask to be used for all IP addresses."
+                    return
+                }
+            }
+
+            if (Test-Bound -Not -ParameterName SubnetIP) {
+                # No subnets (subnet IPs) were supplied but we can calculate them with the netmask
+                $SubnetIP = for ($ipIndex = 0 ; $ipIndex -lt $IPAddress.Count ; $ipIndex++) {
+                    ($IPAddress[$ipIndex].Address -band $SubnetMask[$ipIndex].Address) -as [ipaddress]
+                }
+            } else {
+                if ($IPAddress.Count -ne $SubnetIP.Count) {
+                    if ($SubnetIP.Count -eq 1) {
+                        # If one subnet IP is supplied, let's assume we want to use the same subnet for every IP
+                        $SubnetIP = $SubnetIP * $IPAddress.Count
+                    } else {
+                        Stop-Function -Message "When specifying subnet IPs explicitly, the number of subnets must match the number of IPs, or use one subnet to be applied to all IPs."
+                        return
+                    }
+                }
+            }
+        }
+
         foreach ($ag in $InputObject) {
             if ((Test-Bound -Not -ParameterName Name)) {
                 $Name = $ag.Name
             }
-            if ($Pscmdlet.ShouldProcess($ag.Parent.Name, "Adding $($IPAddress.IPAddressToString) to $($ag.Name)")) {
+            if ($Pscmdlet.ShouldProcess($ag.Parent.Name, "Adding $($IPAddress) to $($ag.Name)")) {
```

Comment:
hey brian, why did the IPAddressToString get removed? I recall it being required for the proper output. Do you have a screenshot to share?

Reply:
Strange, it shouldn't have been required previously, because the default stringification is the same, a la:
```
"$([ipaddress]'127.0.0.1')"
```

The reason I took it out was because `$IPAddress` became an array and I was expecting to use the default array-to-string functionality of joining with `$OFS` (space). Which does work, but as I respond to this I realize it should have worked the same without changing it too, since `.IPAddressToString` would just return an array via the PowerShell array member magic, which I just confirmed as well:

```powershell
"$(([ipaddress[]]@('127.0.0.1','127.0.0.2')))"
"$(([ipaddress[]]@('127.0.0.1','127.0.0.2')).IPAddressToString)"
```

So yeah.. the clarity / intent of the original is indeed better!
(added back)